ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,407 @@ let order_f (_, vm_rec) =
+                (fun vm_cnt_to_hosts_map hs ->
+                  match hs with
+                  | (_, vm_cnt) :: _ ->
+                      IntMap.add vm_cnt (List.map fst hs) vm_cnt_to_hosts_map
+                  | [] ->
+                      vm_cnt_to_hosts_map
+                )
+                IntMap.empty
+           )
+           grp_to_ranked_hosts_map
+       )
+       VMGroupMap.empty
+
+(** Update the 2 layers Map - grp_ranked_hosts when a new VM from the anti-affinity group is
+    planned to run on the host. *)
+let update_anti_affinity_grp_ranked_hosts group host grp_ranked_hosts =
+  let ranked_hosts =
+    try VMGroupMap.find group grp_ranked_hosts
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, group not in it" ;
+      raise Not_found
+  in
+  let vm_cnt_on_host =
+    try
+      IntMap.find_first
+        (fun vm_cnt -> List.mem host (IntMap.find vm_cnt ranked_hosts))
+        ranked_hosts
+      |> fst
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, host not in it" ;
+      raise Not_found
+  in
+  VMGroupMap.update group
+    (fun _ ->
+      Some
+        (ranked_hosts
+        |> IntMap.update vm_cnt_on_host (Option.map (List.filter (( <> ) host)))
+        |> IntMap.update (vm_cnt_on_host + 1) (fun hs ->
+               Some (host :: Option.value ~default:[] hs)
+           )
+        )
+    )
+    grp_ranked_hosts
+
+(*****************************************************************************************************)
+(* Planning code follows                                                                             *)
+
+(** Try to get a spread evenly plan for anti-affinity VMs (for each anti-affinity group, the number of 
+    running VMs from that group are spread evenly in all the rest hosts in the pool):
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+    vms_increasing vm_can_boot_on_host grp_ranked_hosts mapping =
+  debug "vm_anti_affinity_spread_evenly_plan" ;
+  log_vms_increasing ~__context vms_increasing ;
+  log_hosts_increasing ~__context hosts_increasing ;
+  log_grp_ranked_hosts ~__context grp_ranked_hosts ;
+  match vms_increasing with
+  | [] ->
+      mapping
+  | (vm, vm_size) :: remaining_vms -> (
+      let host_selector hosts =
+        hosts_increasing
+        |> List.filter (fun (h, size) -> List.mem h hosts && size >= vm_size)
+        |> List.map fst
+        |> fst_host_can_run vm_can_boot_on_host vm
+      in
+      match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
```

Comment:
This can be done outside of the function.

Reply:
Yes, we can filter only anti-affinity VMs before entering this function, but I think then this function is still needed inside this function to get the `anti-affinity group` of a VM.

Reply:
The group can be together within `vm` in `vms`.